### PR TITLE
add report on cells count

### PR DESCRIPTION
**Pull request by arthurjolo**
Fixes [#1695](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/issues/1695)

This PR adds a command (report_cell_count) to report the cells count before CTS, after CTS and after Filler cells have been added.

### CODE REVIEW

File: src/odb/src/db/odb.tcl [hunk 1]
```diff
@@ -465,6 +465,102 @@ proc report_voltage_domains {} {
...
+  set time_repr_buffer "0"
+  
+  utl::report "\nReporting Cells:"
+
+  foreach inst $insts {
+    set master [$inst getMaster]
+    set master_type [$master getType]
+    set source_type [$inst getSourceType]
+    set cell_name [$master getName]
+    if { bool([$master isBlock]) == bool(true) } {
```
Comment (maliberty) [inline src/odb/src/db/odb.tcl:497]: can you write
```
    if { [$master isBlock] } {
```
?

Comment (arthurjolo) [reply src/odb/src/db/odb.tcl:497]: I don't think tcl supports this, but I can try.

Comment (arthurjolo) [reply src/odb/src/db/odb.tcl:497]: Yes, It works, I will change for the other cases as well.

**Comment by QuantamHD** [comment]
Should we be writing this in TCL? It seems like we should add something to the C++ api that returns a map<cell_type, int> or something like that instead.

**Comment by maliberty** [comment]
I agree this shouldn't be in TCL as it won't be available in python.  You could separate the stats summary, as Ethan suggests, from the actual reporting. 

I don't think the "Cells before/after" messages are needed.

**Comment by arthurjolo** [comment]
A small change now that the code is in C++, the command to report the cells count is now: `sta::report_cell_count`

File: src/dbSta/src/dbSta.cc [hunk 2]
```diff
@@ -270,6 +271,129 @@ std::set<dbNet*> dbSta::findClkNets(const Clock* clk)
   return clk_nets;
 }
 
+std::map<dbSta::InstType, int> dbSta::getInstancesType()
```
Comment (QuantamHD) [inline src/dbSta/src/dbSta.cc:274]: I suggest change the name of this function countInstancesByType. It's more descriptive of it's actual functionality.

**Comment by QuantamHD** [comment]
> A small change now that the code is in C++, the command to report the cells count is now: `sta::report_cell_count`

Nice! I like the change a lot.

File: src/dbSta/src/dbSta.cc (hunk 2 above)
Comment (arthurjolo) [reply src/dbSta/src/dbSta.cc:274]: I agree, this is a better name.

**Comment by maliberty** [comment]
This looks very close to gui's DbInstDescriptor::getInstanceType & DbInstDescriptor::getInstanceTypeText.  I'd like to avoid having two pieces of code that do nearly, but not quite, the same thing.

Could we move the guts of those functions to your code and call them from GUI?  The count function would iterate and sum by calling this new method per inst.

**Comment by arthurjolo** [comment]
> This looks very close to gui's DbInstDescriptor::getInstanceType & DbInstDescriptor::getInstanceTypeText. I'd like to avoid having two pieces of code that do nearly, but not quite, the same thing.
> 
> Could we move the guts of those functions to your code and call them from GUI? The count function would iterate and sum by calling this new method per inst.

@maliberty I belive this last commit  addresses this. The DbInstDescriptor::getInstanceTypeText I maintained as it was in DbInstDescriptor but using the new dbSta::getInstanceType().

**Comment by maliberty** [comment]
Please show a sample output

**Comment by arthurjolo** [comment]
> Please show a sample output

An output example would be:

Reporting Cells count:
  Macro: 6
  Fill: 523662
  Tapcell: 58169
  Antenna: 412
  Buffer/inverter: 242
  Clock buffer/inverter: 81
  Buffer/inverter from timing repair: 288
  Sequential: 525
  Combinational: 1928

**Comment by precisionmoon** [comment]
> > Please show a sample output
> 
> An output example would be:
> 
> ```
> Reporting Cells count:
>   Macro: 6
>   Fill: 523662
>   Tapcell: 58169
>   Antenna: 412
>   Buffer/inverter: 242
>   Clock buffer/inverter: 81
>   Buffer/inverter from timing repair: 288
>   Sequential: 525
>   Combinational: 1928
> ```

Does "Combinational" include buffers, clock buffers, buffers from timing repair?
The report can be made clearer with something like the following:

Cell usage report:
  Macros: 6
  Fill cells: 523662
  Tap cells: 58169
  Antenna cells: 412
  Data buffers/inverters: 242
  Clock buffers/inverters: 81
  Timing repair data buffers/inverters: 288
  Sequential cells: 525
  Complex combinational cells: 1928

Alternate Tcl command name like report_cell_usage may be good also.

**Comment by arthurjolo** [comment]
> Complex

Combinational do not include buffers/inverters. I changed the report.
That wxample now is:
 
```
Cell usage report:
  Macros: 6
  Fill cells: 523662
  Tap cells: 58169
  Antenna cells: 412
  Buffer/inverters: 242
  Clock buffer/inverters: 81
  Timming Repair Buffer/inverters: 288
  Sequential cells: 525
  Complex combinational cells: 1928
```

**Comment by precisionmoon** [comment]
> > Complex
> 
> Combinational do not include buffers/inverters. I changed the report. That wxample now is:
> 
> ```
> Cell usage report:
>   Macros: 6
>   Fill cells: 523662
>   Tap cells: 58169
>   Antenna cells: 412
>   Buffer/inverters: 242
>   Clock buffer/inverters: 81
>   Timming Repair Buffer/inverters: 288
>   Sequential cells: 525
>   Complex combinational cells: 1928
> ```

Just a small typo:  "Timming" should read "Timing"

**Comment by maliberty** [comment]
Please add a unit test and command documentation.

**Comment by arthurjolo** [comment]
@maliberty I am having some difficulties on finding where to add the documentation for the command. The dbSta directory doesn’t have a README.md file, is there another directory I can add the documentation?

**Review by precisionmoon — APPROVED** [review]

**Review by maliberty — APPROVED** [review]

**Pull request merged**